src/backend/mps/compute.rs

```diff
@@ -18,23 +23,408 @@ impl MpsCompute {
         })
     }
 
-    pub fn dispatch_compute(
+    pub fn create_buffer<T: Copy>(&self, data: &[T]) -> Result<Buffer, MpsError> {
+        let buffer = self.device.device().new_buffer_with_data(
+            data.as_ptr() as *const _,
+            std::mem::size_of_val(data) as u64,
+            MTLResourceOptions::StorageModeShared,
+        );
+
+        Ok(buffer)
+    }
+
+    // pub fn dispatch_compute(
+    //     &self,
+    //     pipeline: &ComputePipelineState,
+    //     grid_size: MTLSize,
+    //     thread_group_size: MTLSize,
+    // ) -> Result<(), crate::backend::MpsError> {
+    //     let command_buffer = self.command_queue.new_command_buffer();
+    //     let compute_encoder = command_buffer.new_compute_command_encoder();
+
+    //     compute_encoder.set_compute_pipeline_state(pipeline);
+    //     compute_encoder.dispatch_thread_groups(grid_size, thread_group_size);
+    //     compute_encoder.end_encoding();
+
+    //     command_buffer.commit();
+    //     command_buffer.wait_until_completed();
+
+    //     Ok(())
+    // }
+
+    pub fn matmul(
         &self,
-        pipeline: &ComputePipelineState,
-        grid_size: MTLSize,
-        thread_group_size: MTLSize,
-    ) -> Result<(), crate::backend::MpsError> {
+        a: &Buffer,
+        b: &Buffer,
+        m: usize,
+        n: usize,
+        k: usize,
+    ) -> Result<Buffer, MpsError> {
+        if m == 0 || n == 0 || k == 0 {
+            return Err(MpsError::InvalidDimensions);
+        }
+
+        let result_size = m * k * std::mem::size_of::<f32>();
+        let result_buffer = self
+            .device
+            .device()
+            .new_buffer(result_size as u64, MTLResourceOptions::StorageModeShared);
+
+        // Create dimension buffers
+        let m_buffer = self.create_buffer(&[m as u32])?;
+        let n_buffer = self.create_buffer(&[n as u32])?;
+        let k_buffer = self.create_buffer(&[k as u32])?;
+
+        let library = self
+            .device
+            .device()
+            .new_library_with_source(
+                include_str!("../../../shaders/metal/matrix_ops.metal"),
```

Comment:
Is there a way to transfer this section to build.rs too? or set the library from the `new()` and call it from `self`

> Me being Lazy to point out the lines from the start, commenting on every single line XD